### ISSUE TITLE
sputnikvm panicked at 'arithmetic operation overflow'

### ISSUE THREAD

**Issue by sudachen** [ethereumproject/go-ethereum#432]
```test Special/OverflowGasMakeMoney
thread '<unnamed>' panicked at 'arithmetic operation overflow', C:\Home\.cargo\registry\src\github.com-1ecc6299db9ec823\etcommon-bigint-0.2.9\src\uint\mod.rs:1184:0
note: Run with `RUST_BACKTRACE=1` for a backtrace.
Exception 0xe1525354 0x13ac960 0x0 0x7ffe6ce71f28
PC=0x7ffe6ce71f28
signal arrived during external code execution

github.com/sudachen/playground/branch/sputnik/vm/vendor/github.com/ethereumproject/sputnikvm-ffi/go/sputnikvm._Cfunc_sputnikvm_fire(0x13acae0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	github.com/sudachen/playground/branch/sputnik/vm/vendor/github.com/ethereumproject/sputnikvm-ffi/go/sputnikvm/_obj/_cgo_gotypes.go:406 +0x74
github.com/sudachen/playground/branch/sputnik/vm/vendor/github.com/ethereumproject/sputnikvm-ffi/go/sputnikvm.(*VM).Fire.func1(0x13acae0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	C:/Projects/_GoPkg_/src/github.com/sudachen/playground/branch/sputnik/vm/vendor/github.com/ethereumproject/sputnikvm-ffi/go/sputnikvm/sputnikvm.go:412 +0xa2
github.com/sudachen/playground/branch/sputnik/vm/vendor/github.com/ethereumproject/sputnikvm-ffi/go/sputnikvm.(*VM).Fire(0xc0420bc010, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	C:/Projects/_GoPkg_/src/github.com/sudachen/playground/branch/sputnik/vm/vendor/github.com/ethereumproject/sputnikvm-ffi/go/sputnikvm/sputnikvm.go:412 +0x78
github.com/sudachen/playground/branch/sputnik/vm.(*nvm).Execute(0xa43ad8, 0xc04225e000, 0xc042262000, 0x89f880, 0xc0420ca050, 0x673153c197862a8e, 0xc00bbf6e7e, 0x30, 0xc042037b78, 0x48f469, ...)
	C:/Projects/_GoPkg_/src/github.com/sudachen/playground/branch/sputnik/vm/vm.go:102 +0x526
github.com/sudachen/playground/tests.RunStateTest(0xc0420a80f0, 0xc0420d08d0, 0xc0421d61c0, 0x14, 0xa24480, 0x89c800, 0xa43ad8, 0x22, 0x0)
	C:/Projects/_GoPkg_/src/github.com/sudachen/playground/tests/classic_statetests.go:64 +0x2ae
github.com/sudachen/playground/tests.RunStateTests(0xc0420a80f0, 0xc042037f38, 0x8b2ac0)
	C:/Projects/_GoPkg_/src/github.com/sudachen/playground/tests/classic_statetests.go:111 +0x5c2
github.com/sudachen/playground/branch/sputnik/tests.TestStateSpecial.func2(0xc0420a80f0)
	C:/Projects/_GoPkg_/src/github.com/sudachen/playground/branch/sputnik/tests/classic_state_test.go:30 +0xdd
testing.tRunner(0xc0420a80f0, 0x642d38)
	C:/Go/src/testing/testing.go:746 +0xd7
created by testing.(*T).Run
	C:/Go/src/testing/testing.go:789 +0x2e5

goroutine 1 [chan receive]:
testing.(*T).Run(0xc04208a000, 0x638070, 0x10, 0x642d40, 0x481501)
	C:/Go/src/testing/testing.go:790 +0x303
testing.runTests.func1(0xc04208a000)
	C:/Go/src/testing/testing.go:1004 +0x6b
testing.tRunner(0xc04208a000, 0xc04205dde0)
	C:/Go/src/testing/testing.go:746 +0xd7
testing.runTests(0xc042003700, 0x8b2c00, 0x2, 0x2, 0x0)
	C:/Go/src/testing/testing.go:1002 +0x2df
testing.(*M).Run(0xc04205df18, 0xc04205df70)
	C:/Go/src/testing/testing.go:921 +0x118
main.main()
	github.com/sudachen/playground/branch/sputnik/tests/_test/_testmain.go:46 +0xe2

goroutine 5 [chan receive]:
testing.tRunner.func1(0xc04208a0f0)
	C:/Go/src/testing/testing.go:722 +0x138
testing.tRunner(0xc04208a0f0, 0x642d40)
	C:/Go/src/testing/testing.go:748 +0xe1
created by testing.(*T).Run
	C:/Go/src/testing/testing.go:789 +0x2e5

goroutine 18 [sleep]:
time.Sleep(0x12a05f200)
	C:/Go/src/runtime/time.go:65 +0x145
github.com/sudachen/playground/branch/sputnik/tests.TestStateSpecial.func1(0xc0420a8000)
	C:/Projects/_GoPkg_/src/github.com/sudachen/playground/branch/sputnik/tests/classic_state_test.go:24 +0x45
testing.tRunner(0xc0420a8000, 0x642d30)
	C:/Go/src/testing/testing.go:746 +0xd7
created by testing.(*T).Run
	C:/Go/src/testing/testing.go:789 +0x2e5
rax     0x673153c197862a8e
rbx     0x345fd90
rcx     0x74534fa90bbf6e7e
rdi     0xc0420370c0
rsi     0x13ac960
rbp     0xc042037080
rsp     0x345fb00
r8      0x68a1c4c212a33fd3
r9      0xffffffffffffffff
r10     0x0
r11     0x1
r12     0x0
r13     0x20
r14     0x12
r15     0x4
rip     0x7ffe6ce71f28
rflags  0x202
cs      0x33
fs      0x53
gs      0x2b
```

```javascript
    "OverflowGasMakeMoney" : {
        "env" : {
            "currentCoinbase" : "2adc25665018aa1fe0e6bc666dac8fc2697ff9ba",
            "currentDifficulty" : "0x02b8feb0",
            "currentGasLimit" : "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
            "currentNumber" : "0x00",
            "currentTimestamp" : "0x01",
            "previousHash" : "5e20a0453cecd065ea59c37ac63e079ee08998b6045136a8ce6635c7912ec0b6"
        },
        "logs" : [
        ],
        "out" : "0x",
        "post" : {
            "a94f5374fce5edbc8e2a8697c15331677e6ebf0b" : {
                "balance" : "0x03e8",
                "code" : "0x",
                "nonce" : "0x00",
                "storage" : {
                }
            }
        },
        "postStateRoot" : "a60566e0ecd43f9224e59c41de05e376869357327052aba6a61614fbcccf32ac",
        "pre" : {
            "a94f5374fce5edbc8e2a8697c15331677e6ebf0b" : {
                "balance" : "0x03e8",
                "code" : "0x",
                "nonce" : "0x00",
                "storage" : {
                }
            }
        },
        "transaction" : {
            "data" : "",
            "gasLimit" : "0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe0b",
            "gasPrice" : "0x01",
            "nonce" : "0x00",
            "secretKey" : "45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8",
            "to" : "b94f5374fce5edbc8e2a8697c15331677e6ebf0b",
            "value" : "0x01f5"
        }
    },
```

**Comment by sorpaas** [comment]
See https://github.com/ethereumproject/sputnikvm/issues/298. This currently looks like the expected behavior. If `go-ethereum` passes an invalid transaction to SputnikVM, I don't think there's much can be done. Correct me if I'm wrong.

**Comment by sudachen** [comment]
We must change this behavior to return the error. It is obvious that a client's failure on an incorrect transaction is unacceptable. It does not matter whether such a transaction can occur in a real blockchain or not. This is a potential DoS attack point even if we do not know how to use it now.

**Comment by sorpaas** [comment]
@sudachen Can you double check? `go-ethereum` should handle the transaction validation before passing it to the transaction processor. So in real life there's no chance SputnikVM receives any invalid transactions.

**Comment by sudachen** [comment]
Real live is very complex. There is a potential point of failure related to consensus. If someone will found the way to pass go-client's checking for transaction which can be semi-correct and exactly incorrect for VM?!  

We see there is a way to crash VM (and whole client). I think it's unacceptable. However it's my opinion of cause, but I strongly advise you to change this behavior.

**Comment by sorpaas** [comment]
I'm not talking about philosophy. By "real life" I mean attack through network interface and/or local RPC endpoints. If you find a way to get it crashed there, it would be much appreciated and I would be more than glad to fix it.

What you mentioned is indeed on the roadmap, but won't get fixed until v0.10 due to interface change.

**Comment by sudachen** [comment]
In philosophy program cant panic by input data, only by algorithmic errors.  I am telling about exactly real life. In real life as you see developer of part A can think that developer of part B handles all potential errors and makes all checks, however developer of part B can forget/pass something or think that this checks will be done by developer of part A. IT IS REAL LIFE.  And of cause "philosophically" what does VM is very bad way to handle errors. However it is not so dramatically as real life post-effects.

By the way, since it is in the roadmap it's good.

**Comment by sorpaas** [comment]
Okay thanks!

**tldr**: this is an incorrect library usage. The correct behavior is panic. In sputnikvm v0.10 we will have better error message, but won't change the panic behavior.

Close as we use https://github.com/ethereumproject/sputnikvm/issues/298 to track this issue.
